src/mpid/ch4/src/ch4_recv.h

```diff
@@ -136,19 +139,32 @@ MPL_STATIC_INLINE_PREFIX int MPID_Recv_init(void *buf,
             MPIR_ERR_POP(mpi_errno);
         }
 
-        mpi_errno =
-            MPIDI_NM_mpi_recv_init(buf, count, datatype, rank, tag, comm, context_offset, av,
-                                   &(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request)));
+        MPIDI_CH4I_REQUEST(*request, is_local) = 1;
 
-        if (mpi_errno != MPI_SUCCESS) {
-            MPIR_ERR_POP(mpi_errno);
-        }
+        if (!MPIR_Request_is_complete(*request) && !MPIDI_CH4U_REQUEST_IN_PROGRESS(*request)) {
+            mpi_errno =
+                MPIDI_NM_mpi_recv_init(buf, count, datatype, rank, tag, comm, context_offset, av,
+                                       &(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request)));
 
-        MPIDI_CH4I_REQUEST(*request, is_local) = 1;
-        MPIDI_CH4I_REQUEST(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request), is_local) = 0;
+            if (mpi_errno != MPI_SUCCESS) {
+                MPIR_ERR_POP(mpi_errno);
+            }
+
+            MPIDI_CH4I_REQUEST(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request), is_local) = 0;
+
+            /* cancel the shm request if netmod/am handles the request from unexpected queue. */
+            if (MPIR_Request_is_complete(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request))) {
```

Comment:
Same argument here: I don't think it makes sense to check for request completion here. (I would make sense for regular receive, but remember this line is for persistent request.)

Reply:
I think this one is easier since the request will never be complete, we can just remove this code.

Reply:
Yes, that's what I would do.

Reply:
It is not true. 
This line for the netmod request MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request).
Netmod request can be completed after the call MPIDI_NM_mpi_irecv(buf, count, datatype, rank, tag, comm, context_offset, av, &(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request)));

Reply:
@SergeyGubanov - I think we're talking about the persistent call, `MPID_Recv_init`, rather than the regular receive, `MPID_Recv`.

`MPIDI_NM_mpi_irecv` can definitely complete the request, but I don't know that `MPIDI_NM_mpi_recv_init` is allowed to carry out the full receive operation, particularly since the user hasn't yet called `MPI_Start` on it.

As @hajimefu mentioned above, the request cannot be completed at this point (you only call `MPI_Recv_init` once so having the request be freed shouldn't be a problem). In fact, I'm wondering if the call to `MPIDI_CH4U_REQUEST_IN_PROGRESS` is necessary as the request can't be in progress at this point either. This is really just to set up the request object.

Reply:
@wesbland @SergeyGubanov Yeah, so to summarize, this kind of logic is necessary but it should be in `MPID_Startall`, not here (`MPID_Recv_init`). I happen to be working with `ANY_SOURCE` receive in persistent communication and now fixing it in `MPID_Startall`. I'll upstream the patch soon.